tenant/repository_test.go

```diff
@@ -143,6 +146,162 @@ func (s *TenantServiceTestSuite) TestLookupTenantByNamespace() {
+		svc := tenant.NewDBService(s.DB)
+		mappedVersions := testdoubles.GetMappedVersions(environment.DefaultEnvTypes...)
+
+		// when
+		firstBatch, err := svc.GetTenantsToUpdate(mappedVersions, 5, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, firstBatch, 5)
+		assertContentOfTenants(t, firstBatch, fxt.Tenants, true)
+		updateAllTenants(t, firstBatch, svc, false)
+
+		// and when
+		secondBatch, err := svc.GetTenantsToUpdate(mappedVersions, 5, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, secondBatch, 5)
+		assertContentOfTenants(t, secondBatch, fxt.Tenants, true)
+		assertContentOfTenants(t, secondBatch, firstBatch, false)
+		updateAllTenants(t, secondBatch, svc, true)
+
+		// and when
+		thirdBatch, err := svc.GetTenantsToUpdate(mappedVersions, 5, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, thirdBatch, 1)
+		assertContentOfTenants(t, thirdBatch, fxt.Tenants, true)
+		assertContentOfTenants(t, thirdBatch, firstBatch, false)
+		assertContentOfTenants(t, thirdBatch, secondBatch, false)
+		updateAllTenants(t, thirdBatch, svc, false)
+
+		// and when
+		lastBatch, err := svc.GetTenantsToUpdate(mappedVersions, 5, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, lastBatch, 0)
+	})
+}
+
+func updateAllTenants(t *testing.T, toUpdate []*tenant.Tenant, svc tenant.Service, failed bool) {
+	mappedVersions := testdoubles.GetMappedVersions(environment.DefaultEnvTypes...)
+	for _, tnnt := range toUpdate {
+		namespaces, err := svc.GetNamespaces(tnnt.ID)
+		assert.NoError(t, err)
+		for _, ns := range namespaces {
+			if failed {
+				ns.State = "failed"
+			} else {
+				ns.Version = mappedVersions[string(ns.Type)]
+				ns.State = "ready"
+			}
+			ns.UpdatedBy = "xyz"
+			assert.NoError(t, svc.SaveNamespace(ns))
+		}
+	}
+}
+
+func assertContentOfTenants(t *testing.T, expectedTenants []*tenant.Tenant, slice []*tenant.Tenant, shouldContain bool) {
+	for _, tnnt := range expectedTenants {
+		found := false
+		for _, t := range slice {
+			if t.ID == tnnt.ID {
+				found = true
+				break
+			}
+		}
+		if shouldContain {
+			assert.Truef(t, found, "the slice %s does not contain tenant %s", slice, tnnt)
+		} else {
+			assert.False(t, found, "the slice %s should not contain tenant %s", slice, tnnt)
+		}
+	}
+}
+
+func (s *TenantServiceTestSuite) TestGetSubsetOfFailedTenantsToUpdate() {
+	s.T().Run("returns only those tenants whose namespaces have different updated_by", func(t *testing.T) {
+		// given
+		testdoubles.SetTemplateVersions()
+		controller.Commit = "123abc"
+		previouslyFailed := tf.FillDB(t, s.DB, 1, false, "failed", environment.DefaultEnvTypes...)
+		controller.Commit = "234bcd"
+		tf.FillDB(t, s.DB, 6, false, "failed", environment.DefaultEnvTypes...)
+
+		svc := tenant.NewDBService(s.DB)
+
+		// when
+		result, err := svc.GetTenantsToUpdate(testdoubles.GetMappedVersions(environment.DefaultEnvTypes...), 10, "234bcd")
+
+		// then
+		assert.NoError(t, err)
```

Comment:
same here: use `require.NoError`

Reply:
lazy assertion

Reply:
sorry, I'm not following you

Reply:
sorry, I didn't want to write lazy, but soft assertion :-).
I just like the concept when you let all assertions evaluate and then get all failure messages together.